File: docs/src/testers/functions/batch_invocation.py
# @sniptest filename=batch_invocation.py
# @sniptest show=6-26
# ruff: noqa: E402 -- hidden live fixture precedes displayed imports
import os

function_id = os.environ["NOTTE_FUNCTION_ID"]

from concurrent.futures import ThreadPoolExecutor

from notte_sdk import NotteClient

client = NotteClient()

function = client.Function(function_id=function_id)

urls = ["https://site1.com", "https://site2.com", "https://site3.com"]


def invoke_function(url):
    return function.run(url=url)


# Run in parallel
with ThreadPoolExecutor(max_workers=3) as executor:
    results = list(executor.map(invoke_function, urls))

for result in results:
    print(result.result)

assert all(result.status == "closed" for result in results)
assert [result.result for result in results] == [{"url": url, "search_query": ""} for url in urls]
assert len({result.function_run_id for result in results}) == len(urls)
results = [result.result for result in results]

File: docs/src/testers/functions/batch_invocation.ts
// @sniptest filename=batch_invocation.ts
// @sniptest show=5-22
import assert from 'node:assert/strict';

const functionId = process.env.NOTTE_FUNCTION_ID!;

import { NotteClient } from 'notte-sdk';

const client = new NotteClient();

const fn = client.NotteFunction({ function_id: functionId });

const urls = ['https://site1.com', 'https://site2.com', 'https://site3.com'];

async function invokeFunction(url: string) {
  return fn.run({ url });
}

// Run in parallel
const results = await Promise.all(urls.map(invokeFunction));

for (const result of results) {
  console.log(result.result);
}

assert.ok(results.every(result => result.status === 'closed'));
assert.deepEqual(results.map(result => result.result), urls.map(url => ({ url, search_query: '' })));
assert.equal(new Set(results.map(result => result.function_run_id)).size, urls.length);
const capturedResults = results.map(result => result.result);
export { capturedResults as results };
